[SPARK-49095][SQL] Update `DecimalType`and `decimal fields` compatible logic of `Avro` data source to avoid loss of decimal precision

wayneguow opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR aims to enhance comparison logic to avoid precision loss in decimal parts of `Decimal` in Avro data source. It refers to the related logic of convert `Decimal` type in `Parquet` data source #44513 .

Before:
- As long as the length of the integer part(`precision - scale`) matches, it can be converted.

After:
- Both the integer(`precision - scale`) and decimal part(`scale`) lengths must match.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Fixed the issue causing missing data accuracy.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, stricter matching requirements for conversions between Spark `DecimalType` and `Avro` Decimal type.

Previously, decimal(12,10) -> decimal(5,3) was allowed, but there would be some loss of precision in the decimal part, such as: 13.1234567890 -> 13.123
After that, the exception `avroIncompatibleReadError` will be thrown, because 3 is not greater than or equal 10. Unless both the integer and the decimal part are greater than or equal. For example, decimal(15, 13) is OK because 13>=10 and 15-13>=12-10

But users can restore the legacy behavior, set `spark.sql.legacy.avro.allowIncompatibleDecimalType` to `true`.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Pass GA and add new test cases.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No.

wayneguow commented:
cc @cloud-fan

wayneguow commented:
cc @cloud-fan @LuciferYang  when you have time :-)

wayneguow commented:
> Can we put more information into `Does this PR introduce any user-facing change`? Especially what kind of queries will be broken after this change.

Updated it.

allisonwang-db commented:
@wayneguow, This seems to be a breaking change since the previous working query will now throw exceptions. I would recommend adding a config (by default disabled) and migration docs.

wayneguow committed:
update

wayneguow commented:
> @wayneguow, This seems to be a breaking change since the previous working query will now throw exceptions. I would recommend adding a config (by default disabled) and migration docs.

@allisonwang-db Thanks for your suggestions. I added a new legacy config that users can restore the old behavior. Also cc @cloud-fan
